PR: https://github.com/odoo/odoo/pull/221915

From: 3411f30f6f1b5e5b4fb9ccf7ccdf689caf8b4699
From: Traina Ludo
Date: 2025-12-17 10:59:01

Breaking data model changes scores: del:2 + add:0, change matches:
-    set_tip_after_payment = fields.Boolean('Set Tip After Payment', help="Adjust the amount authorized by payment terminals to add a tip after the customers left or at the end of the day.")
-    pos_set_tip_after_payment = fields.Boolean(compute='_compute_pos_set_tip_after_payment', store=True, readonly=False)

Total Changes: 27

[IMP] pos,pos_discount, ...: tip/discount support fixed or percent value

..., pos_restaurant, pos_adyen, pos_restaurant_adyen, pos_stripe, pos_restaurant_stripe
Task: [#4841174](https://www.odoo.com/odoo/1373/tasks/4841174)

*--

To make tips and discounts more flexible, this commit introduces the ability to choose between a fixed amount or a percentage of the order total.

Additionally, the `set_tip_after_payment` setting was moved from the `pos_restaurant` module to the `point_of_sale` module so it can be used by all POS configs, not just restaurants.
This change impacts other restaurant-related modules that rely on this setting or the tip logic.

closes odoo/odoo#221915

Related: odoo/upgrade#8177
Signed-off-by: Ludo Traina (ltra) <ltra@odoo.com>

================================= pseudo patch: =================================

--- a/addons/pos_restaurant/models/__init__.py
+++ b/addons/pos_restaurant/models/__init__.py
@@ -1,4 +1,3 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from . import pos_config

--- a/addons/pos_restaurant/models/pos_config.py
+++ b/addons/pos_restaurant/models/pos_config.py
@@ -1,7 +1,6 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import api, fields, models, _
+from odoo import _, api, fields, models
 from odoo.tools import convert
 
 
@@ -11,12 +10,11 @@ class PosConfig(models.Model):
     iface_splitbill = fields.Boolean(string='Bill Splitting', help='Enables Bill Splitting in the Point of Sale.')
     iface_printbill = fields.Boolean(string='Bill Printing', help='Allows to print the Bill before payment.')
     floor_ids = fields.Many2many('restaurant.floor', string='Restaurant Floors', help='The restaurant floors served by this point of sale.', copy=False)
-    set_tip_after_payment = fields.Boolean('Set Tip After Payment', help="Adjust the amount authorized by payment terminals to add a tip after the customers left or at the end of the day.")
     default_screen = fields.Selection([('tables', 'Tables'), ('register', 'Register')], string='Default Screen', default='tables')
     use_course_allocation = fields.Boolean(string="Enable Course Allocation")
 
     def _get_forbidden_change_fields(self):
-        forbidden_keys = super(PosConfig, self)._get_forbidden_change_fields()
+        forbidden_keys = super()._get_forbidden_change_fields()
         forbidden_keys.append('floor_ids')
         return forbidden_keys
 
@@ -31,8 +29,6 @@ class PosConfig(models.Model):
                     vals['show_product_images'] = False
                 if 'show_category_images' not in vals:
                     vals['show_category_images'] = False
-            if not is_restaurant or not vals.get('iface_tipproduct', False):
-                vals['set_tip_after_payment'] = False
         pos_configs = super().create(vals_list)
         for config in pos_configs:
             if config.module_pos_restaurant:
@@ -43,10 +39,7 @@ class PosConfig(models.Model):
         if ('module_pos_restaurant' in vals and vals['module_pos_restaurant'] is False):
             vals['floor_ids'] = [(5, 0, 0)]
 
-        if ('module_pos_restaurant' in vals and not vals['module_pos_restaurant']) or ('iface_tipproduct' in vals and not vals['iface_tipproduct']):
-            vals['set_tip_after_payment'] = False
-
-        if ('module_pos_restaurant' in vals and vals['module_pos_restaurant']):
+        if vals.get('module_pos_restaurant'):
             self._setup_default_floor(self)
 
         return super().write(vals)
@@ -149,10 +142,6 @@ class PosConfig(models.Model):
             convert.convert_file(self._env_with_clean_context(), 'pos_restaurant', 'data/scenarios/restaurant_demo_session.xml', idref=None, mode='init', noupdate=True)
         return {'config_id': config.id}
 
-    @api.depends('set_tip_after_payment')
-    def _compute_local_data_integrity(self):
-       super()._compute_local_data_integrity()
-
     def _load_restaurant_demo_data(self, with_demo_data=True):
         self.ensure_one()
         convert.convert_file(self._env_with_clean_context(), 'pos_restaurant', 'data/scenarios/restaurant_category_data.xml', idref=None, mode='init', noupdate=True)

--- a/addons/pos_restaurant/models/res_config_settings.py
+++ b/addons/pos_restaurant/models/res_config_settings.py
@@ -9,7 +9,6 @@ class ResConfigSettings(models.TransientModel):
     pos_floor_ids = fields.Many2many(related='pos_config_id.floor_ids', readonly=False)
     pos_iface_printbill = fields.Boolean(compute='_compute_pos_module_pos_restaurant', store=True, readonly=False)
     pos_iface_splitbill = fields.Boolean(compute='_compute_pos_module_pos_restaurant', store=True, readonly=False)
-    pos_set_tip_after_payment = fields.Boolean(compute='_compute_pos_set_tip_after_payment', store=True, readonly=False)
     pos_default_screen = fields.Selection(related="pos_config_id.default_screen", readonly=False)
     pos_use_course_allocation = fields.Boolean(related='pos_config_id.use_course_allocation', readonly=False)
 
@@ -26,11 +25,3 @@ class ResConfigSettings(models.TransientModel):
                     'pos_iface_printbill': res_config.pos_config_id.iface_printbill,
                     'pos_iface_splitbill': res_config.pos_config_id.iface_splitbill,
                 })
-
-    @api.depends('pos_iface_tipproduct', 'pos_config_id')
-    def _compute_pos_set_tip_after_payment(self):
-        for res_config in self:
-            if res_config.pos_iface_tipproduct:
-                res_config.pos_set_tip_after_payment = res_config.pos_config_id.set_tip_after_payment
-            else:
-                res_config.pos_set_tip_after_payment = False
